flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -95,60 +95,78 @@ public StreamStateHandle registerReference(
-                }
+                // no further handling
+                return entry.stateHandle;
+
+            } else if (entry.stateHandle == newHandle) {
+                // might be a bug but state backend is not required to use a place-holder
+                LOG.debug(
+                        "Duplicated registration under key {} with the same object: {}",
+                        registrationKey,
+                        newHandle);
+            } else if (Objects.equals(entry.stateHandle, newHandle)) {
+                // might be a bug but state backend is not required to use a place-holder
+                LOG.debug(
+                        "Duplicated registration under key {} with the new object: {}.",
+                        registrationKey,
+                        newHandle);
+            } else if (isPlaceholder(newHandle)) {
                 LOG.trace(
-                        "Updating last checkpoint for {} from {} to {}",
+                        "Duplicated registration under key {} with a placeholder (normal case)",
+                        registrationKey);
+                scheduledStateDeletion = newHandle;
+            } else if (entry.confirmed) {
+                LOG.info(
+                        "Duplicated registration under key {} of a new state: {}. "
+                                + "This might happen if checkpoint confirmation was delayed and state backend re-uploaded the state. "
+                                + "Discarding the new state and keeping the old one which is included into a completed checkpoint",
                         registrationKey,
-                        entry.lastUsedCheckpointID,
-                        checkpointID);
-                entry.advanceLastUsingCheckpointID(checkpointID);
-                if (preventDiscardingCreatedCheckpoint) {
-                    entry.preventDiscardingCreatedCheckpoint();
-                }
+                        newHandle);
+                scheduledStateDeletion = newHandle;
+            } else {
+                // Old entry is not in a confirmed checkpoint yet, and the new one differs.
+                // This might result from (omitted KG range here for simplicity):
+                // 1. Flink recovers from a failure using a checkpoint 1
+                // 2. State Backend is initialized to UID xyz and a set of SST: { 01.sst }
+                // 3. JM triggers checkpoint 2
+                // 4. TM sends handle: "xyz-002.sst"; JM registers it under "xyz-002.sst"
+                // 5. TM crashes; everything is repeated from (2)
+                // 6. TM recovers from CP 1 again: backend UID "xyz", SST { 01.sst }
+                // 7. JM triggers checkpoint 3
+                // 8. TM sends NEW state "xyz-002.sst"
+                // 9. JM discards it as duplicate
+                // 10. checkpoint completes, but a wrong SST file is used
+                // So we use a new entry and discard the old one:
+                LOG.info(
```

Comment:
I don't think so, correct me if I am wrong.
Under normal circumstances(not failover), the concurrent checkpoints may share one SST, But except one handle is SST, the others are placeholder (covered by L122), what do you think?

Reply:
We only use placeholder state handle if the checkpoint contains sst has been confirmed

Reply:
Aha, I missed this point, thanks for your clarification. 

Since every upload will give a [new fileName](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java#L442) for SST,  is this(delete the old incompleted checkpoint's statehandle) a bug?

And I accidentally deleted the branch of this PR, I will create a new one to changelog the log level.

Reply:
> Since every upload will give a [new fileName](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/state/filesystem/FsCheckpointStreamFactory.java#L442) for SST, is this(delete the old incompleted checkpoint's statehandle) a bug?

I don't think so, If the checkpoint did not complete successfully, we can't reuse the sst file in it, and the status of the file in the remote file system is unknown (the generated file name is for the file in the remote file system).

Another open question here is whether we need to change the registry's key in incremental checkpoint from the local file name([1][2]) to the remote file name.

[1] https://github.com/apache/flink/blob/c84c9d71dfc577588d788241de230b9619c5b0ef/flink-runtime/src/main/java/org/apache/flink/runtime/state/IncrementalRemoteKeyedStateHandle.java#L327
[2] https://github.com/apache/flink/blob/c84c9d71dfc577588d788241de230b9619c5b0ef/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksIncrementalSnapshotStrategy.java#L397

Reply:
> does this message needs to be info level(or it can be debug level)

I think the probability of reaching this branch is low, so using info level doesn't bring much overhead.

> If the checkpoint did not complete successfully, we can't reuse the sst file in it, and the status of the file in the remote file system is unknown (the generated file name is for the file in the remote file system)

When the number of concurrent checkpoint is bigger than 1, is only the last triggered checkpoint completed?
Suppose`chk-x` and `chk-(x+1)` are pending at the same time, if the state handles of `chk-x` are deleted by the registration of `chk-(x+1)`, `chk-x` is broken.  So, I think this is maybe a bug.

And I open [PR21136](https://github.com/apache/flink/pull/21136) to change the log level, please help to review. :)

Reply:
Sorry for the late reply, I add a UT to verify the concurrent checkpoints logic. https://github.com/fredia/flink/tree/test_concurrent_chk
> the later complete checkpoint will reuse the previous state handle, the logic is [here](https://github.com/apache/flink/blob/8e16cc8e424e352c5b45b46f1520ecf0edec70be/flink-runtime/src/main/java/org/apache/flink/runtime/state/IncrementalRemoteKeyedStateHandle.java#L345)

I found that the previous state handle is discarded when registering a new state handle with same key ([line 260](https://github.com/apache/flink/compare/master...fredia:flink:test_concurrent_chk#diff-80eee7531d8a093be7d2f4cb69dce47453e22e898662c930fb7557908efbebc4R260)), so the later complete checkpoint would use its **own** state handles not previous handles.  
The previous checkpoint's state handles are discarded, although all of the checkpoints may succeed, the succeed checkpoints may be broken. So I think this is a bug.

Reply:
> did not confirm the logic for changelog state backend now, could you please check this logic and update it?

If Changelog state backend delegate Rocksdb state backend, this wrong logic still exist.


I will try to add an ITCase to mock a whole checkpoint and restore scenario, If it is really like the analysis, I will create a ticket.

Reply:
I added an ITCase at https://github.com/fredia/flink/tree/test_concurrent_chk to reproduce this scenario, and created https://issues.apache.org/jira/browse/FLINK-29913 to track this problem.